### ISSUE TITLE
default login/registration web page

### ISSUE THREAD

**Issue by arunkumar413** [FusionAuth/fusionauth-issues#65]
### Problem
The fusionauth doesn't come with a default login page like aws cognito. 

### Solution
Can you also build a simple login/registration page so that it'll help users to speed up prototyping the apps and authentication.

**Comment by robotdan** [comment]
Hi @arunkumar413, FusionAuth does come with a login page.  

If you use the Authorize endpoint in an OAuth2 / OpenID Connect grant workflow you'll be using our login page. https://fusionauth.io/docs/v1/tech/oauth/endpoints#authorize

When you log into FusionAuth itself, you're using the login page. This same login page can be used for all applications. It can also be customized or skinned. https://fusionauth.io/docs/v1/tech/tutorials/login-theme

We do not currently offer a registration builder, so you will need to build the registration page yourself. 

@camilodelvasto Let me see if I can point you in the right direction. 

> However, I'm struggling to get users to sign up using this page, I don't know if I'm missing a parameter or if the URL is different.... has anyone figured this out?

Are you trying to get the URL built to redirect users to the login page? I think what you'll want to do is use the Authorize endpoint. 

For example, if you have your application running on localhost at `https://localhost:8080` and you create an application in FusionAuth with an Id of `4cfc477b-51d7-4f10-b4d5-86970c6b9eba`. 
https://fusionauth.io/docs/v1/tech/oauth/overview

You would then redirect the user to the following URL (with all parameters URL escaped):
`/oauth2/authorize?client_id=4cfc477b-51d7-4f10-b4d5-86970c6b9eba&redirect_uri=https://localhost:8080/login&response_type=code`

The redirect URL must be pre-configured in the FusionAuth application under the OAuth tab. In this example I've used the redirect of `https://localhost:8080/login`.  This URL must handle the rest of the OAuth2 grant flow. 

Let me know if you're still stuck or I didn't answer your question.

**Comment by camilodelvasto** [comment]
Thanks, @robotdan for the fastest response ever!

I was able to use the login page in that way and wanted to register people using the same screen, but I think this:
> We do not currently offer a registration builder, so you will need to build the registration page yourself.
answers my question :(
so you are saying that I need to build my own registration page? I was hoping to be able to use the same social integration to allow people to register... do you happen to have code samples? Also, can I customize the login page to include these features, or should I host it somewhere else and just access it using the API?

Thanks in advance.

**Comment by robotdan** [comment]
Hi @camilodelvasto, correct, you'll need to register new users using the Register API. 

However, if you are going to use Social login, Facebook for example - you do not need to manually register users. When you configure the Facebook Identity Provider, you will enable the application you wish to use it for, and ensure you have enabled `Create registration` toggle is enabled. 
https://fusionauth.io/docs/v1/tech/identity-providers/facebook#create-a-facebook-identity-provider

This will cause the user to be created and registered with any configured default roles if they user does not yet exist, or is not yet registered for the application. 

If you wish to add additional registration data or modify the user roles you can do that using the Update Registration API at any time. 

> .. can I customize the login page to include these features, or should I host it somewhere else and just access it using the API?

I think you're asking if you could add a button to register, or perhaps even add an additional form that supports registration to the login page? Yes, this is possible using the login them, you'd modify the `/oauth2/authorize` template  and add additional forms. 
 https://fusionauth.io/docs/v1/tech/tutorials/login-theme

The form would have to hit your own code though, or you could capture the form submit in JavaScript and then make an API call to some sort of back end server that can securely call the User Registration API.  The customized login page will be shown for all applications however, so if you only want it to show up for one application, you could use an `[#if]` block around the form based upon the `client_id`. 

Will that work for you?

**Comment by camilodelvasto** [comment]
@robotdan Thanks for the detailed info, I'll try what you suggested.

In the meantime, I have another question: on the subject of creating a user using the API, I guess the API would allow me to use my own form from any of my apps, but I wonder about this that I read [here](https://auth0.com/docs/guides/login/universal-vs-embedded):

>(With web apps) the most secure way is to use a central service so the cookies are from the same origin. With embedded login, you'd have to collect the user credentials in an application served from one origin and then send them to another origin, which can present certain security vulnerabilities, including the possibility of a phishing attack.

I was wondering if by using the API to register or log user in are we introducing vulnerabilities. Or maybe I do not understand this completely...

**Comment by robotdan** [comment]
You are correct, the most ideal scenario is to never collect the user's credentials yourself, and always delegate to the identity provider. 

One way for you to approach building your own registration is to not collect a password when you register a new user. If you set the `sendSetPasswordEmail` to `true` when creating the user, the password field will then not be required, and will be ignored if provided. This then will create the user in a state where they cannot log in until they have set up a password. Creating the user in this manner will kick off the Setup Password email to the user and allow them to set their own password. 

https://fusionauth.io/docs/v1/tech/apis/users#create-a-user
https://fusionauth.io/docs/v1/tech/apis/registrations#create-a-user-and-registration-combined

Will this option work for you? 

Another option we've kicked around is - until we have a full registration builder option, we could provide a very simple registration option with a link on the login page to register. This would perhaps just collect email, password, first and last name, and be enabled per application by enabling anonymous registration. In this type of scenario, the user could self-register, and then if you require additional information to complete the registration, you'd collect the additional information at a later time and update their registration.

**Comment by camilodelvasto** [comment]
@robotdan thanks for the tip on creating users without collecting the password. I still need to test this, though, but might be a workaround.

On this subject:
> until we have a full registration builder option

does this mean that you are working on this? Is there an ETA? This would be a really nice addition and critical for us (and I suspect a lot of people) to adopt the service. Keep up the great work!

**Comment by robotdan** [comment]
@camilodelvasto  and @Garogat Yes - we have something in development that is nearly complete. 

It is not a full registration builder, but we are going to deliver something that we hope will satisfy the requirement for most. What we are currently thinking is that we'll add a button to the login page for registration that will be available based upon your configuration. For example, you'll need to enable anonymous registration for a particular application in FusionAuth. 

Here is a high level idea of what will be coming:
1. New registration button added to the login page, enabled through configuration. 
2. Registration page will take email, name, birthday, password, and other top level fields on our User object. Fields may be enabled or disabled via configuration. Email and password will always be required as a minimum. 
3. We will handle the following use cases:
   - User does not exist, create user and register them with the default roles. (registration template)
   - User already exist, create a new registration for the user with the default roles. (complete registration template)


These two new templates will be customizable in the same way the current login pages are using the Theme feature in the UI. 

If your application requires additional registration details such as favorite color, custom application preferences, or something like that - you may need to detect that in your application and have them add additional details that you can then store in FusionAuth using the Update Registration API. 

As mentioned before, this same workflow happens today if you use Social login such as Facebook or Google and enable the 'Registration' toggle in the Identity Provider configuration. In this scenario we create and register the user from the details provided by the external Identity Provider. 

Please let me know if you have any feedback on this plan, and if this proposed feature will adequately cover your use case. 

Thanks to both of you for your use of FusionAuth and for all of the feedback you have provided us. We appreciate your involvement in the FusionAuth community.

**Comment by Garogat** [comment]
@robotdan 
For 99% of my use cases a simple registration as you described it (email, password and a name) as a new user or a application registration for an existing user will be exactly what we need. I guess it will be using the same calls as the api registration does, so already known features like email verification will still work? Thank you again for providing us such a great tool.

**Comment by camilodelvasto** [comment]
Thanks a lot for the update, @robotdan —to my untrained eyes, the plan seems good. Apart from custom fields, I don't think of anything else at the moment that I would need in my workflow.

A couple of things I want to mention:
1. Last week, I sent a message through the contact us form on the website but I haven't got an answer yet. I understand your monetization strategy is based on the support/hosting plans and dev services, and we are very interested in knowing more about what's offered by you, so just a heads up in case you didn't receive it.
2. Thanks so much for creating this product. I believe this is a killer one and much needed in the CIAM space, as other self-hosted options look too cumbersome and are more difficult to install (using docker I was up and running with a test server in minutes!). The one thing I would ask is if you have considered open-sourcing this in the future, to have more people testing, helping, etc.

Thanks again for the great work!

**Comment by arunkumar413** [comment]
@robotdan,

I'm a bit confused. My express/nuxtjs app is running on localhost:3000 and the fusionauth is on localhost:9011

How should I configure below  fusionauth app settings, hyperlink to login page and the port in the authorise URL

Logout URL: 
Authorized origins:  
Authorized redirects:


`
<a href = "http://localhost:PORT/oauth2/authorize?client_id={123456}&redirect_uri=https://localhost:PORT/login&response_type=code"> login </a>
`

**Comment by robotdan** [comment]
Hi @arunkumar413 The redirect to your application will be to `http://localhost:3000` or `https://localhost:3000` depending upon if you've set up local SSL or not. This URL will then need to be authorized in the FusionAuth OAuth configuration.

Are you still having problems with this redirect configuration? 

The registration option is available in version 1.4.0, releasing today. 
https://fusionauth.io/docs/v1/tech/release-notes/

**Comment by arunkumar413** [comment]
@robotdan Yes,  I mean after redirecting, how should I capture the auth code, token and the user details. At the redirected URL page should I send the auth code to fusion to get the token and the user details?
